styx-scheduler-service/src/main/java/com/spotify/styx/Scheduler.java

```diff
@@ -245,44 +244,19 @@ private void limitAndDequeue(Map<String, Resource> resources,
         .filter(resourceRef -> !resources.containsKey(resourceRef))
         .collect(toSet());
 
-    final Set<String> depletedResources = instanceResourceRefs.stream()
-        .filter(resourceId -> {
-          if (!resources.containsKey(resourceId)) {
-            return false;
-          }
-
-          final Resource resource = resources.get(resourceId);
-          final long usage = currentResourceUsage.getOrDefault(resourceId, 0L);
-          return usage >= resource.concurrency();
-        })
-        .collect(toSet());
-
     if (!unknownResources.isEmpty()) {
       stateManager.receiveIgnoreClosed(
           Event.runError(instanceState.workflowInstance(),
               String.format("Referenced resources not found: %s", unknownResources)),
           instanceState.runState().counter());
-    } else if (!depletedResources.isEmpty()) {
-      final Message message = Message.info(
-          String.format("Resource limit reached for: %s",
-              depletedResources.stream()
-                  .map(resources::get)
-                  // Sort resource descriptions to get deterministic message contents
-                  .map(Resource::toString)
-                  .sorted()
-                  .collect(toList())));
-      if (!instanceState.runState().data().message().map(message::equals).orElse(false)) {
-        stateManager.receiveIgnoreClosed(Event.info(instanceState.workflowInstance(), message),
-            instanceState.runState().counter());
-      }
     } else {
       double sleepingTime = dequeueRateLimiter.acquire();
       if (sleepingTime > 0.0001) {
         LOG.debug("Dequeue rate limited and slept for {} ms", sleepingTime * 1000);
       }
 
-      instanceResourceRefs.forEach(id -> currentResourceUsage.computeIfAbsent(id, id_ -> 0L));
-      instanceResourceRefs.forEach(id -> currentResourceUsage.compute(id, (id_, l) -> l + 1));
+      // Racy: some resources may have been removed (become unknown) by now; in that case the
```

Comment:
Is this really a race _issue_? Resources that are removed should after all not be used to limit executions anymore...

Reply:
I think it's not dangerous, but UX wise it's an issue anyway I think because resources that are removed normally cause workflow to error out (it references a deleted resource in its configuration - bad), instead of silently running without limiting.

Reply:
Oh I get what you mean now. Deleting resources doesn't happen very often, we can live with this for now.